File: Algorithms/search_sort/C/mergesort.c
#include <stdlib.h>
#include <stdio.h>

int* merge(int*, int*, int, int, int, int);

int* sort(int*, int, int);

int main(){

    int n;
    scanf("%d", &n);
    int* arr = malloc(sizeof(int) * n);
    for(int i = 0; i < n; i++)
        scanf("%d", arr + i);
    int* x = sort(arr, 0, n - 1);
    for(int i = 0; i < n; i++)
        printf("%d ", *(x + i));
    free(arr);
    free(x);
    return 0;
}



int* merge(int* A, int* B, int A_b, int A_e, int B_b, int B_e){

    int n = A_e - A_b + 1, m = B_e - B_b + 1;
    int i = A_b, j = B_b, k = 0;
    int* final = malloc(sizeof(int) * (n + m));
    // while(i <= A_e && j <= B_e){
    //     if(*(A + i) < *(B + j)){
    //         *(final + k) = *(A + i);
    //         ++i;
    //     }
    //     else{
    //         *(final + k) = *(B + j);
    //         ++j;
    //     }
    //     ++k;
    // }
    while(i <= A_e && j <= B_e){
        if(A[i] < B[j]){
            final[k] = A[i];
            ++i;
        }
        else{
            final[k] = B[j];
            ++j;
        }
        ++k;
    }
    for(;i <= A_e; i++, ++k)
        final[k] = A[i];
    for(;j <= B_e; j++, ++k)
        final[k] = B[j];
    return final;
}



int* sort(int* arr, int B, int E){

    if(B == E)
        return arr;
    else if(B > E)
        return NULL;
    int mid = (B + E) / 2;
    int* left = sort(arr, B, mid);
    int* right = sort(arr, mid + 1, E);
    return merge(left, right, B, mid, mid + 1, E);
}

File: Algorithms/search_sort/Python/mergesort.py
def merge(left_array, right_array):
    i, j, n, m = 0, 0, len(left_array), len(right_array)
    merged_array = []
    while i < n and j < m:
        if left_array[i] <= right_array[j]:
            merged_array.append(left_array[i])
            i += 1
        else:
            merged_array.append(right_array[j])
            j += 1
    merged_array.extend(left_array[i:])
    merged_array.extend(right_array[j:])
    return merged_array


def sort(array):
    n = len(array)
    if n <= 1:
        return array

    left_array = sort(array[:n // 2])
    right_array = sort(array[n // 2:])

    array = merge(left_array, right_array)
    return array
